src/transferslot.cpp

```diff
@@ -68,59 +84,76 @@ TransferSlot::TransferSlot(Transfer* ctransfer)
-
-                    startpos = endpos;
-                    endpos = ChunkedHash::chunkceil(startpos, transfer->size);
+                    while (!asyncIO[i]->finished)
+                    {
+                        transfer->client->waiter->wait();
+                    }
 
-                    bufstart += chunksize;
-                    buflen -= chunksize;
-                    cachetransfer = true;
+                    if (!asyncIO[i]->failed)
+                    {
+                        LOG_verbose << "Async write succeeded";
+                        HttpReqDL *downloadRequest = (HttpReqDL *)reqs[i];
+                        for (chunkmac_map::iterator it = downloadRequest->chunkmacs.begin(); it != downloadRequest->chunkmacs.end(); it++)
+                        {
+                            transfer->chunkmacs[it->first] = it->second;
+                        }
+                        downloadRequest->chunkmacs.clear();
+                        transfer->progresscompleted += downloadRequest->bufpos;
+                        LOG_debug << "Cached async data at: " << downloadRequest->dlpos << "   Size: " << downloadRequest->bufpos;
+                        cachetransfer = true;
+                    }
                 }
+            }
 
-                if (buflen >= SymmCipher::BLOCKSIZE)
-                {
-                    ChunkMAC &chunk = transfer->chunkmacs[ChunkedHash::chunkfloor(startpos)];
-                    transfer->key.ctr_crypt(bufstart, buflen, startpos, transfer->ctriv,
-                                chunk.mac, false, !chunk.finished && !chunk.offset);
+            // Open the file in synchonous mode
+            fa->closef();
+            fa->fopen(&transfer->localfilename, false, true);
+        }
 
-                    fa->fwrite((const byte*)bufstart, buflen, startpos);
-                    chunk.offset += (unsigned int)buflen;
+        for (int i = 0; i < connections; i++)
+        {
+            HttpReqDL *downloadRequest = (HttpReqDL *)reqs[i];
+            if (fa && downloadRequest && downloadRequest->status == REQ_INFLIGHT
+                    && downloadRequest->contentlength == downloadRequest->size
+                    && downloadRequest->bufpos >= SymmCipher::BLOCKSIZE)
+            {
+                downloadRequest->finalize(transfer);
+                m_off_t dlpos = downloadRequest->dlpos;
+                m_off_t bufsize = downloadRequest->bufpos;
+                if (fa->fwrite(downloadRequest->buf, bufsize, dlpos))
+                {
+                    LOG_verbose << "Sync write succeeded";
+                    for (chunkmac_map::iterator it = downloadRequest->chunkmacs.begin(); it != downloadRequest->chunkmacs.end(); it++)
+                    {
+                        transfer->chunkmacs[it->first] = it->second;
+                    }
+                    downloadRequest->chunkmacs.clear();
+                    transfer->progresscompleted += bufsize;
```

Comment:
The write above seems to be also wrong, despite it shouldn't have any consequence because the chunkmap is correctly managed in finalize(). I think that the most appropriate is to modify `bufpos` inside finalize(), because it's the function that manages the encryption. That way, this code would be automatically right.